src/algebra/big_operators/basic.lean

```diff
@@ -531,17 +540,15 @@ calc (∏ x in s, f x) = ∏ x in s, 1 : finset.prod_congr rfl h
   ... = 1 : finset.prod_const_one
 
 @[to_additive] lemma prod_apply_dite {s : finset α} {p : α → Prop} {hp : decidable_pred p}
-  (f : Π (x : α), p x → γ) (g : Π (x : α), ¬p x → γ) (h : γ → β) :
+  {hp' : decidable_pred (λ x, ¬ p x)} (f : Π (x : α), p x → γ) (g : Π (x : α), ¬p x → γ)
```

Comment:
why is `hp'` here, especially as a non-TC argument?

Reply:
1. `hp'` is here for `filter (λ x, ¬ p x)`. In some cases (e.g., if we reuse some construction that relies on `classical`), we have a `filter` or `dite` in a formula that uses classical decidability for `¬ p x`, and it is not the same as `decidable.not` on top of the classical decidability for `p x`.
2. with non-TC arguments we reuse whatever `decidable` instances we have in the formula, no matter what do we have in the environment.

Reply:
Doesn't this mean that rewriting in the forward direction leaves Lean unable to find this goal by unification, creating an annoying goal that needs `apply_instance` to close?

Reply:
@eric-wieser You're right
```suggestion
  [decidable_pred (λ x, ¬ p x)] (f : Π (x : α), p x → γ) (g : Π (x : α), ¬p x → γ)
```